src/sync/app.hpp

```diff
@@ -0,0 +1,121 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or utilied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_APP_HPP
+#define REALM_APP_HPP
+
+#include "sync_user.hpp"
+#include "app_credentials.hpp"
+#include "generic_network_transport.hpp"
+
+namespace realm {
+namespace app {
+
+#pragma mark RealmApp
+/**
+ The `RealmApp` has the fundamental set of methods for communicating with a MongoDB
+ Realm application backend.
+
+ This class provides access to login and authentication.
+
+ Using `serviceClient`, you can retrieve services, including the `RemoteMongoClient` for reading
+ and writing on the database. To create a `RemoteMongoClient`, pass `remoteMongoClientFactory`
+ into `serviceClient(fromFactory:withName)`.
+
+ You can also use it to execute [Functions](https://docs.mongodb.com/stitch/functions/).
+
+ Finally, its `RLMPush` object can register the current user for push notifications.
+
+ - SeeAlso:
+ `RemoteMongoClient`,
+ `RLMPush`,
+ [Functions](https://docs.mongodb.com/stitch/functions/)
+ */
+class App {
+public:
+    struct Config {
```

Comment:
Currently in SDK's we also support setting a custom auth header for proxies and any other custom headers. I suspect defining them here would make sense. But I don't know if that is relevant if we only are going to support MongoDB Realm (compared to selfhosted) as a backend in an initial release?

Reply:
We do not need to worry about self hosted right now.